# OCIAD1 Analysis :dna:

Repository content &#x1F333; &#x1F4C1;

```
â”‚   GO_vs_peroxisomeDB.csv
â”‚   README.md
â”‚
â”œâ”€â”€â”€data
â”‚   â”‚   Human.MitoCarta3.0.csv
â”‚   â”‚   peroxisomeDB.csv
â”‚   â”‚   proteinGroups.txt
â”‚   â”‚
â”‚   â”œâ”€â”€â”€cleaned
â”‚   â””â”€â”€â”€GO_enrichment
â”‚
â”œâ”€â”€â”€lipidomics
â”‚       Lipid_analysis_final.Rmd
â”‚
â”œâ”€â”€â”€proteomics
â”‚   â””â”€â”€â”€code
â”‚       â”‚   EigenMS.zip
â”‚       â”‚   fold_change_calculation_MITOS.R
â”‚       â”‚   fold_change_calculation_TOTALS.R
â”‚       â”‚   fxn.R - self-written functions
â”‚       â”‚   GO_enrichment.R
â”‚       â”‚   imputation_MITOS.R - imputation using the Ludovic method
â”‚       â”‚   imputation_TOTALS.R - imputation using the Ludovic method
â”‚       â”‚   normalization_MITOS.R - normalization using the EigenMS method
â”‚       â”‚   normalization_TOTALS.R - normalization using the EigenMS method
â”‚       â”‚   peroxisomeDB.R
â”‚       â”‚   plotting.R
â”‚       â”‚   supplementary_excel_file.R
â”‚       â”‚
â”‚       â””â”€â”€â”€EigenMS
â”‚
â””â”€â”€â”€supplementary_files
```

This repository contains the statistical analysis of the distributions of the mass spectrometry samples. 
To see the result you should clone this repository to your own local machine, extract the `EigenMS.zip` 
folder, set the working directory to `OCIAD1-Analysis` and then run the files.

## Proteomics analysis

1. Run the `imputation_MITOS.R` and `imputation_TOTALS.R` files first.
2. Next is the normalization, so run the files `normalization_MITOS.R` and `normalization_TOTALS.R`.
3. The last process is the fold change calculation. For this you have to run the files `fold_change_calculation_MITOS.R` and `fold_change_calculation_TOTALS.R`.
4. Now you can also analyse the data in terms of gene ontology (GO term analysis). To do this, run the `GO_enrichment.R` file.
5. The data is ready for plotting using the `plotting.R` file.
6. You can also create a supplementary file with all the data using the `supplementary_excel_file.R` file.

Good luck! ðŸ˜‰

## Lipidomics analysis

Open and run the `Lipid_analysis_final.Rmd` file.

Enjoy!

Authors of code: [Vanessa Linke](https://github.com/vanilink), [Mateusz Chodkowski](https://github.com/matiich), [Kacper Kaszuba](https://github.com/KacperKaszuba0608)

## References

1) PMID: 19602524. "Normalization of peak intensities in bottom-up MS-based proteomics using singular value decomposition".
Karpievitch YV, Taverner T, Adkins JN, Callister SJ, Anderson GA, Smith RD, Dabney AR.
Bioinformatics 2009

2) "Metabolomics data normalization with EigenMS"
Karpievitch YV, Nikolic SB, Wilson R, Sharman JE, Edwards LM.
PLoS One 2014
